guard/src/rules/eval_context.rs

```diff
@@ -1132,23 +1207,154 @@ impl<'value, 'loc: 'value> EvalContext<'value, 'loc> for RootScope<'value, 'loc>
         self.scope
             .resolved_variables
             .insert(variable_name, result.clone());
-        return Ok(result);
+        Ok(result)
     }
 
     fn add_variable_capture_key(
         &mut self,
         variable_name: &'value str,
-        key: &'value PathAwareValue,
+        key: Rc<PathAwareValue>,
     ) -> Result<()> {
         self.scope
             .resolved_variables
             .entry(variable_name)
             .or_default()
-            .push(QueryResult::Resolved(key));
+            .push(QueryResult::Resolved(Rc::clone(&key)));
         Ok(())
     }
 }
 
+pub(crate) fn validate_number_of_params(name: &str, num_args: usize) -> Result<()> {
+    match name {
+        "join" => {
+            if num_args != 2 {
+                return Err(Error::ParseError(format!(
+                    "join function requires 2 arguments be passed, but received {num_args}"
+                )));
+            }
+        }
+        "substring" | "regex_replace" => {
+            if num_args != 3 {
+                return Err(Error::ParseError(format!(
+                    "{name} function requires 3 arguments to be passed, but received {num_args}"
+                )));
+            }
+        }
+        "count" | "json_parse" | "to_upper" | "to_lower" | "url_decode" => {
+            if num_args != 1 {
+                return Err(Error::ParseError(format!(
+                    "{name} function requires 1 argument to be passed, but received {num_args}"
+                )));
+            }
+        }
+        _ => {
+            return Err(Error::ParseError(format!(
+                "no such function named {name} exists"
+            )))
+        }
+    }
+
+    Ok(())
+}
+
+pub(crate) fn try_handle_function_call(
+    fn_name: &str,
+    args: &[Vec<QueryResult>],
+) -> Result<Vec<Option<PathAwareValue>>> {
+    let value = match fn_name {
+        "count" => vec![Some(PathAwareValue::Int((
+            Path::root(),
+            count(&args[0]) as i64,
+        )))],
+        "json_parse" => json_parse(&args[0])?,
+        "regex_replace" => {
+            let substring_err_msg = |index| {
+                let arg = match index {
+                    2 => "second",
+                    3 => "third",
+                    _ => unreachable!(),
+                };
+
+                format!("regex_replace function requires the {arg} argument to be a string")
```

Comment:
What happens if multiple args fail validation? Will it show the error message only for the first failure and terminate?